src/diem/jsonrpc/async_client.py

```diff
@@ -156,11 +146,20 @@ def __init__(
         session_factory: typing.Callable[[], ClientSession] = ClientSession,
     ) -> None:
         self._url: str = server_url
-        self._session_factory = session_factory
         self._last_known_server_state: State = State(chain_id=-1, version=-1, timestamp_usecs=-1)
         self._retry: Retry = retry or Retry(DEFAULT_MAX_RETRIES, DEFAULT_RETRY_DELAY, StaleResponseError)
         self._rs: RequestStrategy = rs or RequestStrategy()
         self._logger: Logger = logger or getLogger(__name__)
+        self._session: aiohttp.ClientSession = session_factory()
+
+    async def close(self) -> None:
+        await self._session.close()
+
+    async def __aenter__(self) -> "AsyncClient":
+        return self
+
+    async def __aexit__(self, *args, **kwargs) -> None:  # pyre-ignore
```

Comment:
if you wanted to maintain the same semantics, could you just use `__del__` instead of the async dunder methods?

Reply:
the problem with `__del__` is it is not awaited, close connection need to be inside async event loop.